### PR TITLE
Add Pretty Printer for syntactic trait parameters (#545)

### PR DESCRIPTION
**Pull request by xavier-fernandez**

### CODE REVIEW

File: scalameta/scalameta/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala [hunk 1]
```diff
@@ -178,6 +178,12 @@ class SyntacticSuite extends scala.meta.tests.parsers.ParseSuite {
     assert(tpe("Foo | Bar").show[Syntax] === "Foo | Bar")
   }
 
+  test("trait parameters - Dotty") {
+    val Scala211 = null // TODO: #389
```
Comment (olafurpg) [inline scalameta/scalameta/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala]: This is not necessary, I think. This should work

```scala
@ import dialects.Dotty
@ Dotty(q"trait T(a: Int)").syntax
res15: String = "trait T"
// OR
@ (Dotty, source("trait T(a: Int)")).syntax
```

File: scalameta/trees/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala [hunk 2]
```diff
@@ -414,7 +414,12 @@ object TreeSyntax {
         case t: Defn.Var       => s(w(t.mods, " "), kw("var"), " ", r(t.pats, ", "), t.decltpe, " ", kw("="), " ", t.rhs.map(s(_)).getOrElse(s(kw("_"))))
         case t: Defn.Type      => s(w(t.mods, " "), kw("type"), " ", t.name, t.tparams, " ", kw("="), " ", t.body)
         case t: Defn.Class     => s(w(t.mods, " "), kw("class"), " ", t.name, t.tparams, w(" ", t.ctor, t.ctor.mods.nonEmpty), templ(t.templ))
-        case t: Defn.Trait     => s(w(t.mods, " "), kw("trait"), " ", t.name, t.tparams, templ(t.templ))
+        case t: Defn.Trait     =>
+          if (!dialect.allowTraitParameters && t.ctor.mods.nonEmpty) {
```
Comment (olafurpg) [inline scalameta/trees/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala]: Just minor nitpick, but I try to keep the happy path first

```diff
-          if (!dialect.allowTraitParameters && t.ctor.mods.nonEmpty) { 
+          if (dialect.allowTraitParameters || t.ctor.mods.isEmpty) { 
```

It flows better in my head for some reason, but maybe it's just me :/

File: scalameta/scalameta/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala (hunk 1 above)
Comment (xeno-by) [inline scalameta/scalameta/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala]: We also have `DottySuite.scala` :)

File: scalameta/trees/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala (hunk 2 above)
Comment (xavier-fernandez) [inline scalameta/trees/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala]: 👍

File: scalameta/scalameta/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala (hunk 1 above)
Comment (xavier-fernandez) [inline scalameta/scalameta/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala]: :(
```
[error]  both object Dotty in package dialects of type meta.dialects.Dotty.type
[error]  and object Scala211 in package dialects of type meta.dialects.Scala211.type
```

**Comment by xavier-fernandez** [comment]
@olafurpg @xeno-by amend done.

**Comment by xeno-by** [comment]
LGTM! Thank you for your contribution.

**Pull request merged**